build-support/src/main/kotlin/aws/sdk/kotlin/gradle/crt/CMakeTasks.kt

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.sdk.kotlin.gradle.crt
+
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.tasks.TaskProvider
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
+import org.jetbrains.kotlin.konan.target.HostManager
+import org.jetbrains.kotlin.konan.target.KonanTarget
+
+/**
+ * See [CMAKE_BUILD_TYPE](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html)
+ */
+enum class CMakeBuildType {
+    Debug,
+    RelWithDebInfo,
+    Release,
+}
+
+/**
+ * Configure CMake tasks for building and installing CRT locally for a given Kotlin/Native target.
+ *
+ * This function sets up the following tasks:
+ * * cmakeConfigure<NativeTargetName>   -- e.g. cmakeConfigureLinuxX64
+ * * cmakeBuild<NativeTargetName>       -- e.g. cmakeBuildLinuxX64
+ * * cmakeInstall<NativeTargetName>     -- e.g. cmakeInstallLinuxX64
+ *
+ * CMake tasks may or may not run inside of a docker container depending on the native target being built.
+ * All linux targets are built in a container.
+ *
+ * @param knTarget the native target to build CRT for
+ * @param buildType the [CMakeBuildType] to build for CMake build type. Defaults to `RelWithDebInfo` since end users
+ * can always strip the binary of all debug info.
+ * @return the `cmakeInstall` task for the target which can be used to wire up additional task dependency relationships
+*/
+fun Project.configureCrtCMakeBuild(
+    knTarget: KotlinNativeTarget,
+    buildType: CMakeBuildType = CMakeBuildType.RelWithDebInfo,
+): TaskProvider<Task> {
+    val cmakeConfigure = registerCmakeConfigureTask(knTarget, buildType)
+
+    val cmakeBuild = registerCmakeBuildTask(knTarget, buildType)
+    cmakeBuild.configure {
+        dependsOn(cmakeConfigure)
+    }
+
+    val cmakeInstall = registerCmakeInstallTask(knTarget, buildType)
+    cmakeInstall.configure {
+        dependsOn(cmakeBuild)
+    }
+
+    return cmakeInstall
+}
+
+internal fun Project.registerCmakeConfigureTask(
+    knTarget: KotlinNativeTarget,
+    buildType: CMakeBuildType,
+): TaskProvider<Task> {
```

Comment:
**Question**: Why are `registerCmakeConfigureTask`, `registerCmakeBuildTask`, `registerCmakeInstallTask`, and `runCmake` marked as `internal`? They seem to be used only in this file and thus feel like they could be `private`.

Reply:
sure we can make them private, I was probably just toying around with how to structure this 🤷